os_migrate/plugins/module_utils/resource.py

```diff
@@ -0,0 +1,198 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from ansible_collections.os_migrate.os_migrate.plugins.module_utils import const
+from ansible_collections.os_migrate.os_migrate.plugins.module_utils import exc
+
+
+class Resource():
+
+    resource_type = 'UNDEFINED'
+    sdk_class = 'UNDEFINED'
+
+    # Serialization
+    info_from_sdk = []
+    info_from_refs = []
+    params_from_sdk = []
+    params_from_refs = []
+
+    # Deserialization
+    sdk_params_from_params = []
+    sdk_params_from_refs = []
+
+    # ===== PUBLIC CLASS/STATIC METHODS (alphabetic sort) =====
+
+    # This constructor should work OOTB in all cases, shouldn't be
+    # overriden in child classes.
+    @classmethod
+    def from_data(cls, data):
+        res_type = data.get('type', None)
+        if res_type != cls.resource_type:
+            raise exc.UnexpectedResourceType(cls.resource_type, res_type)
+
+        obj = cls()
+        # Barring resource type errors, we allow Resource subclass
+        # construction with invalid data so that we don't break on
+        # loading a data file, but we explicitly fail with full list
+        # of errors (rather than a single exception) when all
+        # resources are validated with is_data_valid() and/or
+        # data_errors().
+        obj.data = data
+        # Just in case the data didn't contain those keys (invalid
+        # data), set to empty defaults so that all other methods can
+        # rely on these existing rather than calling dict.get().
+        obj.data.setdefault(const.RES_PARAMS, {})
+        obj.data.setdefault(const.RES_INFO, {})
+        return obj
+
+    # Meant to be extended in child classes, but can be overriden
+    # completely.
+    @classmethod
+    def from_sdk(cls, conn, sdk_resource):
+        if not isinstance(sdk_resource, cls.sdk_class):
+            raise exc.UnexpectedResourceType(
+                cls.sdk_class, sdk_resource.__class__)
+
+        obj = cls()
+        obj._data_from_sdk_and_refs(
+            sdk_resource, cls._refs_from_sdk(conn, sdk_resource))
+        obj.data['type'] = cls.resource_type
+        return obj
+
+    # === PRIVATE CLASS/STATIC METHODS (alphabetic sort) ===
+
+    # Used when creating Resource from SDK object, should be overriden
+    # in majority of child classes.
+    @staticmethod
+    def _refs_from_sdk(conn, sdk_res):
+        return {}
+
+    # Not meant to be overriden in majority of subclasses.
+    @classmethod
+    def _set_sdk_params_same_name(cls, ser_params, sdk_params, param_names):
+        """Copy values from `ser_params` into `sdk_params` for all keys in
+        `param_names` (list of strings), only for values in `ser_params`
+        which aren't None.
+        """
+        for p_name in param_names:
+            cls._set_sdk_param(ser_params, p_name, sdk_params, p_name)
+
+    # Not meant to be overriden in majority of subclasses.
+    @staticmethod
```

Comment:
The only difference between static and class methods (line 71, line 81) is whether we need to call another static/class method from within. I'm playing with an idea to make them all `classmethod` for consistency and simply not use the first param in some of them. It probably doesn't matter at all in practice though :).